optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java

```diff
@@ -42,4 +49,36 @@ public static Object wrapProblem(Class<?> solutionClass, OpaquePythonReference p
                     " on a function that return a list of int).", t);
         }
     }
+
+    public static Object wrapFact(Class<?> factClass, OpaquePythonReference fact, Map<Number, Object> referenceMap) {
+        try {
+            // We are wrapping a fact; so setters will not be used
+            PythonObject out = (PythonObject) PythonWrapperGenerator.wrap(factClass, fact,
+                    referenceMap,
+                    PythonWrapperGenerator.NONE_PYTHON_SETTER);
+            out.visitIds(out.get__optapy_reference_map());
```

Comment:
My ignorance will show here, bear with me: what is the reference map, where does it come from, how is it used?

Reply:
The reference map is a map from Python ID to Java Object. it is populated during cloning (https://github.com/optapy/optapy/blob/f2d0143fec8052aae4834a9c2d449b2f75ad686c/optapy-core/src/main/python/optaplanner_java_interop.py#L125-L249) and during object translation (https://github.com/optapy/optapy/blob/f2d0143fec8052aae4834a9c2d449b2f75ad686c/jpyinterpreter/src/main/python/python_to_java_bytecode_translator.py#L194-L300). To prevent memory leaks, the map is a "Map with Extras" (that is, a `Map` that delegates to another map EXCEPT for write operations (i.e. put is not delegated)); the base map (which is used only for `get`) is the `jpyinterpreter` reference map (which store things like constants and types, which need to be the same between different functions). The extra map is populated during cloning, which adds the planning solution, planning entities and planning list variables.